### PR TITLE
Bluetooth: Mesh: Model extension concept

### PR DESCRIPTION
**Pull request by trond-snekvik**
Adds the model extension concept to the access layer, as described in
the Mesh Profile Specification, Section 2.3.6. Extensions are
implemented as a tree, using two pointers in each model:

The extends pointer points to the first extended model, and the next
pointer points to the next sibling or (if the NEXT_IS_PARENT flag is
set) the parent model in the tree, forming a cyclical "Left-child
right-sibling" (LCRS) tree. The tree root can be obtained by calling
bt_mesh_model_root_get(), and the extended models can be walked by
calling bt_mesh_model_tree_walk().

According to the Mesh Profile Specification Section 4.2.3, all models in
the same extension tree share one subscription list per element. This is
implemented by walking the model's extension tree, and pooling the
subscription lists of all models in the same element into one. If the
config server adds a subscription to a model, it may be stored in any of
the model tree's models' subscription lists. No two models in the same
extension tree and element will have duplicate groups listed. This
allows us to increase extended models' capacity for subscriptions
significantly.

Signed-off-by: Trond Einar Snekvik <Trond.Einar.Snekvik@nordicsemi.no>

----

Alternative simplified approach, that just does the sublist sharing, for reference: https://github.com/trond-snekvik/zephyr/commit/52afff31356d1b983bf761554339c04d3e3a6da8

### CODE REVIEW

**Comment by trond-snekvik** [comment]
Looks like this pushes the Microbit mesh application over the RAM limit. @jhedberg, any suggestions for workarounds? I could make this a configurable feature to reduce the memory footprint of each model, but it probably won't be long before this issue pops up again.

File: include/bluetooth/mesh/access.h [hunk 1]
```diff
@@ -495,7 +495,7 @@ struct bt_mesh_model {
 	/* Internal information, mainly for persistent storage */
 	u8_t  elem_idx;   /* Belongs to Nth element */
 	u8_t  mod_idx;    /* Is the Nth model in the element */
-	u16_t flags;      /* Information about what has changed */
+	u16_t flags;      /* Model flags for internal bookkeeping */
```
Comment (joerchan) [inline include/bluetooth/mesh/access.h]: By internal do you mean that the application shouldn't use it?

**Comment by jhedberg** [comment]
>  @jhedberg, any suggestions for workarounds? I could make this a configurable feature to reduce the memory footprint of each model, but it probably won't be long before this issue pops up again.

I'd be fine with adding a Kconfig for this.

File: include/bluetooth/mesh/access.h [hunk 2]
```diff
@@ -511,6 +511,12 @@ struct bt_mesh_model {
 	/* Model callback structure. */
 	const struct bt_mesh_model_cb * const cb;
 
+#ifdef CONFIG_BT_MESH_MODEL_EXTENSIONS
```
Comment (joerchan) [inline include/bluetooth/mesh/access.h]: I think you need to define this for doxygen in order to get it included in the online documentation. But I don't know where exactly. @jhedberg You know this right?

Comment (jhedberg) [reply include/bluetooth/mesh/access.h]: Look for "PREDEFINED" in `doc/zephyr.doxyfile.in`

**Comment by trond-snekvik** [comment]
Kconfig entry and doxygen define added

**Comment by jhedberg** [comment]
@trond-snekvik unless I missed it, it doesn't seem like the new Kconfig option is enabled for any samples, i.e. probably CI isn't exercising the new code. Could you enable it for one or more samples?

**Comment by jhedberg** [comment]
Follow-up to my previous comment - for now it's sufficient that we build test the code. Next it'd be nice to have a sample app showing the use of this API, but that can be done as a later PR.

**Review by jhedberg — APPROVED** [review]

**Pull request merged**